### PR TITLE
add internal and external URL handling for the docker pull secret

### PR DESCRIPTION
**Pull request by deads2k**
This plumbs the information from the master config down to the controllers for internal and external registry values.

@smarterclayton it's small in case you change your mind.
@mfojtik we'll want test coverage on this (beyond my unit test). We need a card?
@bparees I saw your name somewhere

### CODE REVIEW

**Comment by smarterclayton** [comment]
This is super small. 

/approve

**Comment by smarterclayton** [comment]
But needs tests to make sure it doesn't break.

File: pkg/cmd/openshift-controller-manager/conversion.go [hunk 1]
```diff
@@ -25,6 +25,14 @@ func ConvertMasterConfigToOpenshiftControllerConfig(input *configapi.MasterConfi
 	// deep copy to make sure no linger references are shared
 	in := input.DeepCopy()
 
+	registryURLs := []string{}
+	if len(in.ImagePolicyConfig.ExternalRegistryHostname) > 0 {
+		registryURLs = append(registryURLs, in.ImagePolicyConfig.ExternalRegistryHostname)
+	}
+	if len(in.ImagePolicyConfig.InternalRegistryHostname) > 0 {
+		registryURLs = append(registryURLs, in.ImagePolicyConfig.InternalRegistryHostname)
+	}
```
[the remark this answers is not in the document]
Comment (bparees) [reply pkg/cmd/openshift-controller-manager/conversion.go]: k

File: pkg/cmd/server/apis/config/types.go [hunk 2]
```diff
@@ -1574,6 +1575,11 @@ type ServiceAccountControllerConfig struct {
 	ManagedNames []string
 }
 
+type DockerPullSecretControllerConfig struct {
+	// RegistryURLs is a list of urls that the docker pull secrets should be valid for.
+	RegistryURLs []string
```
Comment (mfojtik) [inline pkg/cmd/server/apis/config/types.go]: can we call this AdditionalRegistryURLs to match the controller field?

File: pkg/cmd/openshift-controller-manager/conversion.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mfojtik) [reply pkg/cmd/openshift-controller-manager/conversion.go]: i think OPENSHIFT_DEFAULT_REGISTRY is handled by the logic that sets the ImagePolicyConfig already if I remember correctly

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 3]
```diff
@@ -98,6 +102,9 @@ func NewDockerRegistryServiceController(secrets informers.SecretInformer, servic
 type DockerRegistryServiceController struct {
 	client kclientset.Interface
 
+	// AdditionalRegistryURLs is a list of URLs that are always included
```
Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: nit: lowercase

File: pkg/serviceaccounts/controllers/docker_registry_service.go [hunk 4]
```diff
@@ -218,7 +225,7 @@ func (e *DockerRegistryServiceController) watchForDockerURLChanges() {
 
 // getDockerRegistryLocations returns the dns form and the ip form of the secret
 func (e *DockerRegistryServiceController) getDockerRegistryLocations() []string {
-	ret := []string{}
+	ret := append([]string{}, e.additionalRegistryURLs...)
```
Comment (mfojtik) [inline pkg/serviceaccounts/controllers/docker_registry_service.go]: why not start with ret := e.additionalRegistryURLs ?

File: pkg/cmd/server/apis/config/types.go (hunk 2 above)
Comment (deads2k) [reply pkg/cmd/server/apis/config/types.go]: > nit: optional: can we call this AdditionalRegistryURLs to match the controller field?

I think from the config side, the user see this as his spot to specify the registry urls, not to set extra ones.  From his point of view, we're the ones adding extras.  The API is for them, not us.

File: pkg/serviceaccounts/controllers/docker_registry_service.go (hunk 4 above)
Comment (deads2k) [reply pkg/serviceaccounts/controllers/docker_registry_service.go]: > why not start with ret := e.additionalRegistryURLs ?

fear of accidental mutation.

**Comment by deads2k** [comment]
comments addressed, integration test added.

**Comment by mfojtik** [comment]
@deads2k the integration failure seems real.

**Pull request merged**